subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java

```diff
@@ -247,4 +257,26 @@ public SourceSet resources(Action<? super SourceDirectorySet> configureAction) {
     public SourceDirectorySet getAllSource() {
         return allSource;
     }
+
+    @Override
+    public ExtensionContainer getExtensions() {
+        return getConvention();
+    }
+
+    @Override
+    public DynamicObject getAsDynamicObject() {
+        assertExtensibleDynamicObject();
+        return extensibleDynamicObject;
+    }
+
+    public Convention getConvention() {
+        assertExtensibleDynamicObject();
+        return extensibleDynamicObject.getConvention();
+    }
+
+    private void assertExtensibleDynamicObject() {
+        if (extensibleDynamicObject == null) {
+            extensibleDynamicObject = new ExtensibleDynamicObject(this, SourceSet.class, instantiator);
+        }
+    }
```

Comment:
Why aren't you doing this in the constructor the way that the `ProjectImpl` does?

Also, is there any concern about race conditions from multiple threads instantiating multiple instances of the `ExtensibleDynamicObject` for `this`?

Reply:
Right, let's make it simple first